### ISSUE TITLE
Configure confuse: difference between --advertise-client-urls and --listen-client-urls, etc.

### ISSUE THREAD

**Issue by sirchen** [coreos/etcd#9923]
@gyuho Thanks very much for your help of these days. I think I'm understanding the cert configure gradually. 

I have read the [Configuration](https://github.com/coreos/etcd/blob/6a265731e10a5137b991c1aa3a83ecefdd149d50/Documentation/faq.md#configuration) FAQ. Still one problem is confusing me, namely the difference between `--advertise-client-urls` and `--listen-client-urls` in etcd server. 

Many tests have been done by using etcdctl. It seems that only `--listen-client-urls` is used to decide which **address:port** is used and whether we use tls or not. And the  **address:port** I just mentioned is used by client(etcdctl, gRPC-proxy, etc.) I didn't see any use of `--advertise-client-urls`, which completely seems to be useless.

As for me, The difference between --advertise-client-urls` and `--listen-addr` have the same question also.

**Comment by gyuho** [comment]
`--advertise-client-urls` may be used for proxy peer discovery, so you want to specify that (do not use `localhost` for advertise URLs) if you plan to run several gRPC proxies.

**Comment by sirchen** [comment]
Sorry to disturb again, I still don't understand where the flag `--advertise-client-url` will be used in gRPC-proxy mode.

Server usage:
`ETCDCTL_API=3   ./etcd grpc-proxy start  --cacert=./config/ssl/ca.pem  --cert=./config/ssl/client.pem  --key=./config/ssl/client-key.pem --trusted-ca-file=./config/ssl/ca.pem --cert-file=./config/ssl/client.pem --key-file=./config/ssl/client-key.pem --endpoints xx.xx.xx.xx:2379  --listen-addr xx.xx.xx.xx:23795  --advertise-client-url xx.xx.xx.xx:23780`

Client usage:
`ETCDCTL_API=3   etcdctl  --cacert=./config/ssl/ca.pem  --cert=./config/ssl/client.pem  --key=./config/ssl/client-key.pem --endpoints=[https://10.20.124.33:23795`
When I use member list comand, it shows as follows:


<img width="533" alt="111" src="https://user-images.githubusercontent.com/22660757/42747329-94061a60-890e-11e8-8af6-36af550f1e9e.png">

We know,  without gRPC-proxy, **PEER ADDRS** is used for found peers and **CLIENT ADDRS** is used for client to connect with.

So what can I do with the **CLIENT ADDRS** or how the `--advertise-client-url` and **CLIENT ADDRS**  works?

**Comment by sirchen** [comment]
@gyuho I recently encountered some tough problems and had no idea.

First, gRPC-proxy is not used and application works fine.

Then I use gRPC-proxy, usage is like this:
07.beta is client and 01.beta is server.
`ETCDCTL_API=3   ./etcd grpc-proxy start  --cacert=./config/ssl/ca.pem  --cert=./config/ssl/client.pem  --key=./config/ssl/client-key.pem --trusted-ca-file=./config/ssl/ca.pem --cert-file=./config/ssl/client.pem --key-file=./config/ssl/client-key.pem --endpoints 100.50.129.77:2379  --listen-addr 100.50.129.77:23795  --advertise-client-url 100.50.129.77:23780`

The application connected gRPC-proxy well at first, but then it disconnected from the gRPC-proxy. Using tcpdump to capture the packet, I found this.
![1531819259024-image](https://user-images.githubusercontent.com/22660757/42808405-d7d4ccdc-89e5-11e8-8e52-234aed895f56.png)

My application tried to connect port 23780, and it fails of course. Then everything is out of control, it disconnected from the normal port 23795 and then tried to connect port  23780 again and again.I don't know what happens. 

The flag `--advertise-client-url` completely confuses me now. May I have your suggestions? Thanks in advance.
